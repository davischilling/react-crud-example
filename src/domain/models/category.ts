export interface Category {
  id: string;
  name: string;
  description: string | null;
  is_active: boolean;
  deleted_at: string | null;
  created_at: string;
  updated_at: string;
}
