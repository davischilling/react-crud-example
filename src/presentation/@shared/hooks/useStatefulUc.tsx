import { useCallback, useEffect, useLayoutEffect, useMemo, useRef, useState } from 'react';
import {
  SetState,
  StateCallback,
  StatefulUseCase,
} from '../../../domain/usecases/stateful-usecase';

type StateFulContextUseCaseProps<State, UseCaseClass extends StatefulUseCase<State>> = {
  UseCase: new (state: State, setStateCb: SetState<State>) => UseCaseClass;
  store: () => [State, (value: Partial<State>) => void];
};

function useCallbackRef<T = () => void>(callback: T) {
  const callbackRef = useRef(callback);
  useLayoutEffect(() => {
    callbackRef.current = callback;
  }, [callback]);
  return callbackRef;
}

export function useStatefulUseCase<State, UseCaseClass extends StatefulUseCase<State>>({
  UseCase,
  store,
}: StateFulContextUseCaseProps<State, UseCaseClass>) {
  const [logicState, setLogicState] = store();

  const setState = useCallback(
    (newState: any, cb?: StateCallback) => {
      setLogicState(Object.assign({}, logicState, newState));
      cb && cb();
    },
    [logicState, setLogicState],
  );

  console.log('rendering useStatefulUseCase');

  const useCaseLogicRef = useCallbackRef<UseCaseClass>(
    new UseCase(logicState, (newState, cb) => {
      setState(newState, cb);
    }),
  );

  useEffect(() => {
    return () => {
      useCaseLogicRef.current.init();
    };
  }, []);

  return { state: logicState, useCase: useCaseLogicRef.current };
}
