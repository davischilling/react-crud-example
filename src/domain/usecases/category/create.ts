import { api } from '../../api';
import { ValidationError } from '../../errors/validation-error';
import { Category } from '../../models/category';
import { StatefulUseCase } from '../stateful-usecase';

export type State = {
  category: Partial<Category>;
};

export const DEFAULT_STATE: State = {
  category: {
    name: '',
    description: '',
    is_active: true,
  } as Partial<Category>,
};

export class CreateCategoryUseCase extends StatefulUseCase<State> {
  init = async () => {
    // no need for implementation
  };

  setCategory = ({ name, value }: { name: 'name' | 'description'; value: string }) => {
    if (!['name', 'description'].includes(name)) {
      throw new ValidationError(`Invalid attribute type: ${name}`);
    }
    this.setState({
      category: {
        ...this.state.category,
        [name]: value,
      },
    });
  };

  toggleIsActive = () => {
    this.setState({
      category: {
        ...this.state.category,
        is_active: !this.state.category.is_active,
      },
    });
  };

  createCategory = async () => {
    const fmtCategory: Partial<Category> = {
      ...this.getState().category,
      created_at: new Date().toString(),
      updated_at: new Date().toString(),
      deleted_at: null,
    };
    await api.post('/categories', fmtCategory);
  };
}
