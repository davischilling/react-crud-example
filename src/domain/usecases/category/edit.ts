import { api } from '../../api';
import { Category } from '../../models/category';
import { StatefulUseCase } from '../stateful-usecase';

export type State = {
  isLoading: boolean;
  category: Partial<Category>;
};

export const DEFAULT_STATE: State = {
  isLoading: true,
  category: {
    name: '',
    description: '',
    is_active: true,
  } as Partial<Category>,
};

export class EditCategoryUseCase extends StatefulUseCase<State> {
  public init = async () => {
    await this.initCategory();
  };

  initCategory = async () => {
    const { data } = await api.get(`/categories/${this.state.category.id}`);
    this.setState({ category: data, isLoading: false });
  };

  setCategory = ({ name, value }: { name: 'name' | 'description'; value: string }) => {
    this.setState({
      category: {
        ...this.state.category,
        [name]: value,
      },
    });
  };

  toggleIsActive = () => {
    this.setState({
      category: {
        ...this.state.category,
        is_active: !this.state.category.is_active,
      },
    });
  };

  updateCategory = async () => {
    const fmtCategory: Partial<Category> = {
      ...this.state.category,
      updated_at: new Date().toString(),
    };
    await api.put(`/categories/${this.state.category.id}`, fmtCategory);
  };
}
