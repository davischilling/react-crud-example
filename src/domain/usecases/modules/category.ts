import { api } from '../../api';
import { Category } from '../../models/category';
import { StatefulUseCase } from '../stateful-usecase';

export type State = {
  categories: Category[];
  filteredCategories: Category[];
  category: Partial<Category>;
};

export const DEFAULT_STATE: State = {
  categories: [],
  filteredCategories: [],
  category: {
    name: '',
    description: '',
    is_active: true,
  } as Partial<Category>,
};

export class CategoryUseCase extends StatefulUseCase<State> {
  public init = async () => {
    await this.initCategories();
  };

  initCategories = async () => {
    console.log('initCategories');
    const { data } = await api.get('/categories');
    this.loadCategories(data);
  };

  filterCategories = (search: string | undefined) => {
    const categories = this.getState().categories;
    if (!search) {
      this.setState({ filteredCategories: categories });
      return;
    }
    const filteredCategories = categories.filter((category) => {
      return category.name.toLowerCase().includes(search.toLowerCase());
    });
    this.setState({ filteredCategories });
  };

  setCategory = ({ name, value }: { name: 'name' | 'description'; value: string }) => {
    this.setState({
      category: {
        ...this.getState().category,
        [name]: value,
      },
    });
    // console.log(this.getState().category);

    // console.log({
    //   ...this.state.category,
    //   [name]: value,
    // });

    // this.setState(
    //   {
    //     category: {
    //       ...this.getState().category,
    //       [name]: value,
    //     },
    //   },
    //   () => {
    //     console.log(this.getState().category);
    //   },
    // );
  };

  toggleIsActive = () => {
    this.setState(
      {
        category: {
          ...this.getState().category,
          is_active: !this.getState().category.is_active,
        },
      },
      () => {
        console.log(this.getState().category);
      },
    );
  };

  createCategory = async () => {
    const fmtCategory: Partial<Category> = {
      ...this.getState().category,
      created_at: new Date().toString(),
      updated_at: new Date().toString(),
      deleted_at: null,
    };
    const { data: newCategory } = await api.post('/categories', fmtCategory);
    this.setState({ categories: [...this.state.categories, newCategory] });
  };

  updateCategory = async (category: Partial<Category>) => {
    const fmtCategory: Partial<Category> = {
      ...category,
      updated_at: new Date().toString(),
    };
    const { data: updatedCategory } = await api.put(`/categories/${category.id}`, fmtCategory);
    const categories = this.state.categories.map((category) => {
      if (category.id === updatedCategory.id) {
        return updatedCategory;
      }
      return category;
    });
    this.setState({ categories });
  };

  deleteCategory = async (id: string) => {
    const categories = this.getState().categories;
    const category = categories.find((category) => category.id === id);
    if (!category) {
      throw new Error('Category not found');
    }
    const fmtCategory: Category = {
      ...category,
      deleted_at: new Date().toString(),
    };
    await api.put(`/categories/${id}`, fmtCategory);
    const fmrCategories = categories.map((category) => {
      if (category.id === id) {
        return fmtCategory;
      }
      return category;
    });
    this.loadCategories(fmrCategories);
  };

  private loadCategories = (c: Category[]) => {
    const categories = c.filter((c: Category) => !c.deleted_at);
    this.setState({ categories, filteredCategories: categories });
  };
}
