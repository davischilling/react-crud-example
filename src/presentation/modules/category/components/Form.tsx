import { Box, Button, Grid } from '@mui/material';
import { ChangeEvent, FormEvent } from 'react';
import { Link } from 'react-router-dom';
import { Category } from '../../../../domain/models/category';
import FormInput from '../../../@shared/components/FormInput';
import FormSwitch from '../../../@shared/components/FormSwitch';

type FormProps = {
  category: Partial<Category>;
  isDisabled: boolean;
  isLoading?: boolean;
  handleToggle: (e: ChangeEvent<HTMLInputElement>) => void;
  handleChange: (e: ChangeEvent<HTMLInputElement>) => void;
  handleSubmit: (e: FormEvent<HTMLFormElement>) => Promise<void>;
};

export default function Form({
  category,
  isDisabled,
  handleToggle,
  handleChange,
  handleSubmit,
}: FormProps) {
  return (
    <Box p={2}>
      <form onSubmit={handleSubmit}>
        <Grid container spacing={3}>
          <FormInput
            fullWidth
            required
            name="name"
            label="Name"
            value={category.name}
            onChange={handleChange}
          />
          <FormInput
            fullWidth
            required
            name="description"
            label="Description"
            value={category.description}
            onChange={handleChange}
          />

          <FormSwitch
            label="active"
            name="is_active"
            color="secondary"
            onChange={handleToggle}
            checked={category.is_active}
            inputProps={{ 'aria-label': 'controlled' }}
          />

          <Grid item xs={12}>
            <Box display="flex" gap={2}>
              <Button variant="contained" component={Link} to="/categories">
                Back
              </Button>

              <Button variant="contained" color="secondary" type="submit" disabled={isDisabled}>
                Save
              </Button>
            </Box>
          </Grid>
        </Grid>
      </form>
    </Box>
  );
}
