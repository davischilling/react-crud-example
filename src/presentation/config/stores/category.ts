import { CategoryUseCase, DEFAULT_STATE, State } from '../../../domain/usecases/modules/category';
import { useModuleWrapper } from '../../@shared/hooks/useModuleWrapper';

const { Provider: CategoryProvider, useModuleUseCase: useCategoryUseCase } = useModuleWrapper<
  State,
  CategoryUseCase
>({
  DEFAULT_STATE,
  UseCase: CategoryUseCase,
});

export { CategoryProvider, useCategoryUseCase };
