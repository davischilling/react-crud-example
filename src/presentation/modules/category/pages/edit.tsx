import { Box, Paper, Typography } from '@mui/material';
import { useNavigate, useParams } from 'react-router-dom';

import {
  DEFAULT_STATE,
  EditCategoryUseCase,
  State,
} from '../../../../domain/usecases/category/edit';
import { useFormState } from '../../../@shared/hooks/useFormState';
import { useStatefulUseCase } from '../../../@shared/hooks/useStatefulUc';
import Form from '../components/Form';

export function EditCategoryPage() {
  const id = useParams().id;

  const { state, useCase } = useStatefulUseCase<State, EditCategoryUseCase>({
    UseCase: EditCategoryUseCase,
    DEFAULT_STATE,
    INITIAL_STATE: {
      category: {
        id,
        name: '',
        description: '',
        is_active: true,
      },
    },
  });

  const navigate = useNavigate();

  const { handleChange, handleToggle, handleSubmit } = useFormState({
    setEntity: useCase.setCategory,
    onToggle: useCase.toggleIsActive,
    onSubmit: useCase.updateCategory,
    submitCallback: () => navigate('/categories'),
    snackMessage: 'Category updated successfully',
  });

  return (
    <Box>
      <Paper>
        <Box p={2}>
          <Box mb={2}>
            <Typography variant="h4">Edit Category</Typography>
          </Box>
        </Box>

        <Form
          isLoading={state.isLoading}
          category={state.category}
          isDisabled={state.isLoading}
          handleToggle={handleToggle}
          handleChange={handleChange}
          handleSubmit={handleSubmit}
        />
      </Paper>
    </Box>
  );
}
