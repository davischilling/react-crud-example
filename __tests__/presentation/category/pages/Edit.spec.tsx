import { act, fireEvent, render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter, useNavigate, useParams } from 'react-router-dom';
import { DEFAULT_STATE, EditCategoryUseCase } from '../../../../src/domain/usecases/category/edit';
import * as useStatefulUseCaseModule from '../../../../src/presentation/@shared/hooks/useStatefulUc';
import Form from '../../../../src/presentation/modules/category/components/Form';
import { EditCategoryPage } from '../../../../src/presentation/modules/category/pages/edit';

jest.mock('../../../../src/presentation/modules/category/components/Form', () => ({
  __esModule: true,
  default: jest.fn(({ handleSubmit, handleToggle }) => (
    <div>
      Mocked Form Content
      <form onSubmit={handleSubmit}>
        <button onClick={handleToggle}>toggle</button>
        <button type="submit">submit</button>
      </form>
    </div>
  )),
}));

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
  useParams: jest.fn().mockReturnValue({ id: '1' }),
}));

describe('EditCategoryPage', () => {
  let fragment: DocumentFragment;
  let useStatefulUseCase: jest.SpyInstance;

  let navigateSpy: jest.SpyInstance;
  let updateCategorySpy: jest.SpyInstance;

  let setCategorySpy: jest.SpyInstance;
  let toggleIsActiveSpy: jest.SpyInstance;
  const mockState = { category: { name: 'Mocked Category' }, isLoading: false };

  beforeEach(async () => {
    jest.clearAllMocks();
    navigateSpy = jest.fn();
    (useNavigate as jest.Mock).mockReturnValue(navigateSpy);
    updateCategorySpy = jest.fn();
    setCategorySpy = jest.fn();
    toggleIsActiveSpy = jest.fn();
    useStatefulUseCase = jest
      .spyOn(useStatefulUseCaseModule as any, 'useStatefulUseCase')
      .mockReturnValue({
        state: mockState,
        useCase: {
          updateCategory: updateCategorySpy,
          setCategory: setCategorySpy,
          toggleIsActive: toggleIsActiveSpy,
        },
      });
    fragment = await waitFor(() => {
      const { asFragment } = render(<EditCategoryPage />, { wrapper: BrowserRouter });
      return asFragment();
    });
  });

  test('renders the "Edit Category" page', () => {
    const titleElement = screen.getByText('Edit Category');
    expect(titleElement).toBeInTheDocument();
    expect(fragment).toMatchSnapshot();
  });

  test('calls useStatefulUseCase with correct arguments', () => {
    expect(useStatefulUseCase).toHaveBeenCalledTimes(1);
    expect(useStatefulUseCase).toHaveBeenCalledWith({
      UseCase: EditCategoryUseCase,
      DEFAULT_STATE,
      INITIAL_STATE: {
        category: {
          id: '1',
          name: '',
          description: '',
          is_active: true,
        },
      },
    });
  });

  test('should correctly call useParams', () => {
    expect(useParams).toHaveBeenCalledTimes(1);
    expect(useParams).toHaveBeenCalledWith();
  });

  test('should correctly call useNavigate', () => {
    expect(useNavigate).toHaveBeenCalledTimes(1);
    expect(useNavigate).toHaveBeenCalledWith();
  });

  test('should call navigate and onSubmit when handleSubmit returned by useFormState hook is called', async () => {
    const formSubmitButton = screen.getByText('submit');

    await act(async () => {
      fireEvent.click(formSubmitButton);
    });

    expect(updateCategorySpy).toHaveBeenCalledTimes(1);
    expect(updateCategorySpy).toHaveBeenCalledWith();
    expect(navigateSpy).toHaveBeenCalledTimes(1);
    expect(navigateSpy).toHaveBeenCalledWith('/categories');
  });

  test('passes the correct props to the Form component', () => {
    expect(Form).toHaveBeenCalledWith(
      expect.objectContaining({
        category: mockState.category,
        isDisabled: false,
        handleToggle: expect.any(Function),
        handleChange: expect.any(Function),
        handleSubmit: expect.any(Function),
      }),
      {},
    );
  });
});
