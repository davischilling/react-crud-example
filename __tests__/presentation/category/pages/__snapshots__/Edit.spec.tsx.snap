// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditCategoryPage renders the "Edit Category" page 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
    >
      <div
        class="MuiBox-root css-19midj6"
      >
        <div
          class="MuiBox-root css-1qm1lh"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          >
            Edit Category
          </h4>
        </div>
      </div>
      <div>
        Mocked Form Content
        <form>
          <button>
            toggle
          </button>
          <button
            type="submit"
          >
            submit
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
